### ISSUE TITLE
flux-watch: TypeError: Object of type 'bytes' is not JSON serializable

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#5702]
A user has reported this error and stack trace while using `flux watch`:

```
flux-watch: ERROR: Object of type 'bytes' is not JSON serializable

flux-watch: DEBUG: Traceback (most recent call last):

File "/usr/lib64/flux/python3.6/flux/util.py", line 279, in __call__

main_func()

File "/usr/libexec/flux/cmd/flux-watch.py", line 200, in main

fh.reactor_run()

File "/usr/lib64/flux/python3.6/flux/core/handle.py", line 319, in reactor_run

Flux.raise_if_exception()

File "/usr/lib64/flux/python3.6/flux/core/handle.py", line 133, in raise_if_exception

raise cls.set_exception(None) from None

File "/usr/lib64/flux/python3.6/flux/future.py", line 38, in continuation_callback

py_future.then_cb(py_future, *py_future.then_args, **py_future.then_kwargs)

File "/usr/lib64/flux/python3.6/flux/job/output.py", line 663, in _output_event_watch

self._fulfill_output(event)

File "/usr/lib64/flux/python3.6/flux/job/output.py", line 633, in _fulfill_output

self.fulfill((event.stream, event.render()))

File "/usr/lib64/flux/python3.6/flux/future.py", line 369, in fulfill

result = json.dumps(result).encode("utf-8", errors="surrogateescape")

File "/usr/lib64/python3.6/json/__init__.py", line 231, in dumps

return _default_encoder.encode(obj)

File "/usr/lib64/python3.6/json/encoder.py", line 199, in encode

chunks = self.iterencode(o, _one_shot=True)

File "/usr/lib64/python3.6/json/encoder.py", line 257, in iterencode

return _iterencode(o, 0)

File "/usr/lib64/python3.6/json/encoder.py", line 180, in defaulto.__class__.__name__)
TypeError: Object of type 'bytes' is not JSON serializable

 ```

**Comment by grondo** [comment]
Ah, [base64.b64decode](https://docs.python.org/3/library/base64.html#base64.b64decode) returns `bytes` so that is perhaps the problem. Perhaps this change is needed:

```diff
diff --git a/src/bindings/python/flux/job/output.py b/src/bindings/python/flux/job/output.py
index 9130741bb..1dcb12126 100644
--- a/src/bindings/python/flux/job/output.py
+++ b/src/bindings/python/flux/job/output.py
@@ -113,7 +113,7 @@ class OutputEvent(EventLogEvent):
             data = self.context["data"]
             if "encoding" in self.context:
                 if self.context["encoding"] == "base64":
-                    data = base64.b64decode(data)
+                    data = base64.b64decode(data).encode("utf-8")
             if "repeat" in self.context:
                 data *= self.context["repeat"]
             if labelio:
```

**Comment by grondo** [comment]
The change above alone is not sufficient to fix this issue. First, the `decode` needs to use `errors=surrogateescape` to ensure errors during decoding are handled in some way. Second, by default `sys.stdout` `sys.stderr` and other streams used by `flux watch` are opened with `errors='strict'` by default. This means that if surrogates were used in the decoding of binary data, then writing to these stream with raise another exception. Therefore, streams used in this interface have to be reopened with `errors='surrogateescape'` as well before they can be safely used.

The following change seems to work:

```diff
diff --git a/src/bindings/python/flux/job/output.py b/src/bindings/python/flux/job/output.py
index 9130741bb..2781f1a55 100644
--- a/src/bindings/python/flux/job/output.py
+++ b/src/bindings/python/flux/job/output.py
@@ -113,7 +113,9 @@ class OutputEvent(EventLogEvent):
             data = self.context["data"]
             if "encoding" in self.context:
                 if self.context["encoding"] == "base64":
-                    data = base64.b64decode(data)
+                    data = base64.b64decode(data).decode(
+                        "utf-8", errors="surrogateescape"
+                    )
             if "repeat" in self.context:
                 data *= self.context["repeat"]
             if labelio:
diff --git a/src/bindings/python/flux/job/watcher.py b/src/bindings/python/flux/job/watcher.py
index b8dd5bae6..9e4357277 100644
--- a/src/bindings/python/flux/job/watcher.py
+++ b/src/bindings/python/flux/job/watcher.py
@@ -367,8 +367,8 @@ class JobWatcher:
         self.t0 = starttime
         self.log_events = log_events
         self.log_status = log_status
-        self.stdout = stdout
-        self.stderr = stderr
+        self.stdout = self._reopen(stdout)
+        self.stderr = self._reopen(stderr)
         self.labelio = labelio
         self.exitcode = 0
         self.show_progress = progress
@@ -379,6 +379,22 @@ class JobWatcher:
         if jobs:
             self.add_jobs(*jobs)
 
+    @staticmethod
+    def _reopen(stream):
+        """reconfigure/reopen stream with correct encoding and error handling"""
+        try:
+            # reconfigure() only available in >=3.7
+            stream.reconfigure(encoding="utf-8", errors="surrogateescape")
+            return stream
+        except AttributeError:
+            return open(
+                stream.fileno(),
+                mode="w",
+                encoding="utf-8",
+                errors="surrogateescape",
+                closefd=False,
+            )
+
     @staticmethod
     def _status_to_exitcode(status):
         """Calculate exitcode from job status"""
@@ -420,9 +436,9 @@ class JobWatcher:
         self.progress.add_jobs(*jobs)
 
         if stdout is None:
-            stdout = self.stdout
+            stdout = self._reopen(self.stdout)
         if stderr is None:
-            stderr = self.stderr
+            stderr = self._reopen(self.stderr)
 
         for job in jobs:
             if not self.t0 or job.t_submit < self.t0:

```
